### DISCUSSION TITLE
CMake configuration for USE_CPP_PACKAGE creates incorrect VS 2017 projects

### DISCUSSION THREAD

**Discussion by Serge1985** [apache/incubator-mxnet#13639] [Q&A]
### Description
VS2017 projects generated by cmake with USE_CPP_PACKAGE option fail to build. I had to do a number of modifications to the projects in order to get them to build - ideally it should be incorporated in cmake files.

### Environment info (Required)

C:\Users\schernet>python C:\Users\schernet\source\MXNet\diagnose.py
----------Python Info----------
Version      : 3.6.6
Compiler     : MSC v.1900 64 bit (AMD64)
Build        : ('v3.6.6:4cf1f54eb7', 'Jun 27 2018 03:37:03')
Arch         : ('64bit', 'WindowsPE')
------------Pip Info-----------
Version      : 10.0.1
Directory    : C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python36_64\lib\site-packages\pip
----------MXNet Info-----------
No MXNet installed.
----------System Info----------
Platform     : Windows-10-10.0.17134-SP0
system       : Windows
node         : SCHERNET-2J265
release      : 10
version      : 10.0.17134
----------Hardware Info----------
machine      : AMD64
processor    : Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
Name
Intel(R) Core(TM) i7-7600U CPU @ 2.80GHz

----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.6582 sec, LOAD: 1.8760 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.2683 sec, LOAD: 1.1988 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.2583 sec, LOAD: 1.1738 sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0878 sec, LOAD: 1.2566 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0957 sec, LOAD: 1.9019 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0868 sec, LOAD: 0.5415 sec.

### Build info (Required if built from source)

Visual studio 2017

MXNet commit hash:
f6317c9a92fe2b7dddaecf7891f5ff7f215ae329

Build config:
see below.

## Error Message:

1) In debug configuration, I hit 
error C1083: unabel to open file: “mxnet-cpp/op.h”: No such file or directory.
2) All configurations, gtest* projects have hit
C2220	warning treated as error - no 'object' file generated
Warning	C4996	'std::tr1': warning STL4002: The non-Standard std::tr1 namespace and TR1-only machinery are deprecated and will be REMOVED. You can define _SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING to acknowledge that you have received this warning.	
3) 10 of the projects, e.g. mlp hit:
LNK1181	cannot open input file 'mxnet_static.lib'	

Note, for all these issues, I found a solution by modifying the .vcxproject files - see What have you tried to solve it? below.

## Steps to reproduce

When I cloned git repository, I ran cmake as follows:

`set OpenCV_DIR=c:\Software\Code\Libraries\opencv\build
cmake -G "Visual Studio 15 2017 Win64" .. -T host=x64 -DUSE_CUDA=0 -DUSE_CUDNN=0 -DUSE_OPENCV=1 -DUSE_OPENMP=1 -DUSE_PROFILER=1 -DUSE_MKL_IF_AVAILABLE=1 -DUSE_LAPACK=1 -DUSE_DIST_KVSTORE=0  -DUSE_MKLDNN=0 -DMKL_ROOT=C:/Software/Code/Libraries/IntelSWTools/compilers_and_libraries/windows/mkl -DUSE_CPP_PACKAGE=1  -DUSE_MKLML_MKL=0
-- CMAKE_SYSTEM_NAME Windows
F16C instruction set is not yet supported for MSVC
CMake Warning (dev) at 3rdparty/mshadow/cmake/mshadow.cmake:8 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  CMake variable MKL_ROOT is set to:

    C:/Software/Code/Libraries/IntelSWTools/compilers_and_libraries/windows/mkl

  For compatibility, CMake is ignoring the variable.
Call Stack (most recent call first):
  CMakeLists.txt:289 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found MKL (include: C:/Software/Code/Libraries/IntelSWTools/compilers_and_libraries/windows/mkl/include, lib: C:/Software/Code/Libraries/IntelSWTools/compilers_and_libraries/windows/mkl/lib/intel64/mkl_rt.lib
-- Could NOT find Gperftools (missing: GPERFTOOLS_LIBRARIES GPERFTOOLS_INCLUDE_DIR)
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
-- Could NOT find Jemalloc (missing: JEMALLOC_LIBRARY JEMALLOC_INCLUDE_DIR)
-- OpenCV ARCH: x64
-- OpenCV RUNTIME: vc15
-- OpenCV STATIC: ON
-- Found OpenCV 3.4.1 in C:/Software/Code/Libraries/opencv/build/x64/vc15/lib
-- You might need to add C:\Software\Code\Libraries\opencv\build\x64\vc15\bin to your PATH to be able to run your applications.
--  OpenCV_LIBS=opencv_core;opencv_highgui;opencv_imgproc;opencv_imgcodecs
-- OpenCV 3.4.1 found (C:/Software/Code/Libraries/opencv/build/x64/vc15/lib)
-- Found OpenMP_C: -openmp
-- Found OpenMP_CXX: -openmp
-- Found OpenMP: TRUE
USE_LAPACK is ON
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
-- Could NOT find Jemalloc (missing: JEMALLOC_LIBRARY JEMALLOC_INCLUDE_DIR)
-- Google Test not found
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/schernet/source/MXNet/incubator-mxnet/build`

Then I tried to build Debug/x64 configuration.

## What have you tried to solve it?

1. In my case, since I used opencv, I needed to make **debug** dll (opencv_world341d.dll) available, because libmxnet.dll depends on it:
C:\WINDOWS\system32>mklink  opencv_world341d.dll C:\Software\Code\Libraries\opencv\build\x64\vc15\bin\opencv_world341d.dll
2. For all gtest* projects, I added 4996 to the disabled warnings list
3. mxnet_static.lib needs to be replaced with ..\..\$(Configuration)\libmxnet.lib in all projects (build\cpp-package\example)

I believe 2 and 3 should be addressed by cmake changes.

**Comment by vdantu** [comment]
@mxnet-label-bot add [build, question]
